### ISSUE TITLE
Error: There is no subnet routing to the internet

### ISSUE THREAD

**Issue by BurningDog** [whilenull/7777-support#5]
Thanks for putting this out there, I'm happy to try it out! I'm on Mac, installation was fine. Running it for the first time I get this:

```
$ 7777 --verbose
Using the AWS region us-east-1.
Validating the 7777 license.
Generating unique RSA keys for the SSH tunnel.
Checking if the port 7777 is available.
Port 7777 selected.
Listing databases.
Only one database found in this region, automatically selecting example-dev.
Retrieving the availability zone and subnet of the database.
    Error: There is no subnet routing to the internet in availability zone us-east-1a of VPC vpc-12345678.
```

It's possible I've not set up my VPC correctly - I followed the instructions at https://bref.sh/docs/environment/database.html#accessing-the-internet in order to give my Lambda access to the internet after I moved it into the VPC. This setup worked and my Lambda can access the internet.

In short:

* Created 6 new subnets (1 for each availability zone in us-east-1)
* Created a NAT gateway using the first existing subnet for us-east-1a
  * The route table for that subnet has internet access (I _think_ - because it says Destination `0.0.0.0/0` Target `igw-12345678`
* Created a route table for my VPC
  * In its `Routes` tabs added an entry: `0.0.0.0/0` with the target being the new NAT gateway
  * In its `Subnet Associations` tabs added the 6 new subnets I'd recently created
* Tested my Lambda has internet access, which it now has.

Could it be I messed up with CIDR records? If I look at the IPv4 CIDR entries inside `us-east-1a`, the initial subnet inside the VPC has value `172.31.16.0/20` and the new subnet has value `172.31.96.0/20`.

**Comment by deleugpn** [comment]
Hi @BurningDog, first let me start by saying thank you so much for testing our tool!

I know it's a bit of a complicated subject, but a subnet that points to a NAT gateway is not public, it's actually private. a NAT device can only perform outbound connections and will never allow inbound connections from the internet. What that translates to is that any device using that Subnet will be able to talk to the internet and receive responses via the NAT, but if anyone on the internet tries to initiate a connection with your device, the NAT device will just drop the packets.

For 7777, we absolutely need connection initiated from outside i.e. your computer. So what you're missing is a Route Table that points directly to the `igw-12345678` and a Subnet associated with that Route Table.

Does that make sense?

**Comment by BurningDog** [comment]
Thanks @deleugpn - yes, that does make sense. Thanks for the in-depth explanation!

I've now created a Public Subnet with IPv4 CIDR of `172.31.192.0/20` (thanks to https://network00.com/NetworkTools/IPv4AddressPlanner/) - but turns out I _didn't_ need to edit the `Route Table Association` and link it to the `lamda-to-internet` route table I'd previously created (that results in the incorrect target being the NAT rather than the correct target of the IGW). Then followed step 9 and 10 above to enable auto-assign public IPv4 address.

I tried `7777` again and got this error: `Error: There is no subnet routing to the internet in availability zone us-east-1a of VPC vpc-12345678.`

Ah damn, `7777` is looking for `us-east-1a` but my subnet was created in `us-east-1c` - I didn't specify a specific availability zone when creating the public subnet.

Deleted the subnet, re-created it with the `us-east-1a` availability zone, did all the other things, tried `7777` again, It works!

```
$ 7777
Only one database found in this region, automatically selecting example-dev.
Setting up 7777 in the AWS account, this usually takes 30 seconds.
7777 is now set up, let's connect to the database.

Tunnel created 🎉
Connect to the database on port 7777 on your machine.

Press Ctrl+C to stop and destroy the connection.
```

Thank you!!!

**Comment by BurningDog** [comment]
In order to connect using Sequel Pro on Mac:

![Screenshot 2020-11-06 at 12 53 08](https://user-images.githubusercontent.com/787653/98358287-1f8fc680-202f-11eb-8fd5-dea0a6758e9b.png)

**Comment by mnapoli** [comment]
Awesome, glad it's working for you now!

**Comment by ottaviano** [comment]
hey, I've the same error. Does It mean that for using 7777 we should buy `NAT Gateway` on AWS ($27/month)?

**Comment by mnapoli** [comment]
Hi @ottaviano. No, [in the context of serverless apps] the NAT Gateway is useful specifically:

- with Lambda because Lambda cannot use normal VPC internet gateways to get internet access
- for traffic to go out of the VPC, i.e. so that Lambda can send HTTP requests to the internet

(this is a broad approximation, anyone else feel free to correct me)

NAT Gateway will not change a thing here.

Here your VPC likely has no subnet that is routed (in/out) to the internet. That's needed because you need your computer to connect to the database.

I think [the comment above by Marco](https://github.com/whilenull/7777-support/issues/5#issuecomment-722625995) explains in more details what's probably missing:

> So what you're missing is a Route Table that points directly to the Internet Gateway igw-12345678 and a Subnet associated with that Route Table, so that the Subnet would be allowed to talk to the internet, but also receive data from the internet.

**Comment by deleugpn** [comment]
@ottaviano are you using the default AWS VPC that comes with an AWS Account? I investigated further that VPC and I see that the Subnets that comes by default are not 'explicitly associated' with the Route Table. If you go to **Console -> VPC -> Route Table** you might be able to see the Route Table.
Upon selecting it, there should be a few Tabs on the lower part of the UI. In the **Subnet Association**, you might be able to see no subnet associated, but a note that says that `The following subnets have not been explicitly associated with any route tables and are therefore associated with the main route table:`.

If you click on "Edit Subnet Association" and explicitly associate one of those Subnets to the Route Table, 7777 will be able to use it. So far I have not been able to make 7777 use an "implicitly associated subnet" because I have not found a way to find out which subnet is "implicitly associated".

**Comment by trungpv1601** [comment]
> @ottaviano are you using the default AWS VPC that comes with an AWS Account? I investigated further that VPC and I see that the Subnets that comes by default are not 'explicitly associated' with the Route Table. If you go to **Console -> VPC -> Route Table** you might be able to see the Route Table.
> Upon selecting it, there should be a few Tabs on the lower part of the UI. In the **Subnet Association**, you might be able to see no subnet associated, but a note that says that `The following subnets have not been explicitly associated with any route tables and are therefore associated with the main route table:`.
> 
> If you click on "Edit Subnet Association" and explicitly associate one of those Subnets to the Route Table, 7777 will be able to use it. So far I have not been able to make 7777 use an "implicitly associated subnet" because I have not found a way to find out which subnet is "implicitly associated".

Thank you so much. Save my life 👍

**Comment by matthewsuan** [comment]
@mnapoli @deleugpn 
I have this issue as well although the subnets in question are public. Can you please help?

**Comment by mnapoli** [comment]
@matthewsuan did the comment above help? (it should)

**Comment by matthewsuan** [comment]
@mnapoli yes I was actually able to get around this. thx
